File: queues/recent_counter.py
from collections import deque
from typing import Deque


class RecentCounter:
    def __init__(self):
        self.queue: Deque[int] = deque()

    def ping(self, t: int) -> int:
        self.queue.append(t)
        while self.queue and self.queue[0] < t - 3000:
            self.queue.popleft()
        return len(self.queue)

# TIme Complexity: O(1) for each ping, as each element is added and removed at most once.
# Space Complexity: O(n) in the worst case, where n is the number of pings in the last 3000 milliseconds.

File: queues/RecentCounter.java
import java.util.ArrayDeque;
import java.util.Deque;

public class RecentCounter {
  Deque<Integer> queue;

  public RecentCounter() {
    this.queue = new ArrayDeque<>();
  }

  public int ping(int t) {
    queue.addLast(t);
    while (!queue.isEmpty() && queue.peekFirst() < t - 3000) {
      queue.removeFirst();
    }
    return queue.size();
  }
}

// Time complexity O(n) where n is the number of pings
// Space complexity O(n) in the worst case when all pings are within 3000